### DISCUSSION TITLE
Watch mode not working for custom file extension

### DISCUSSION THREAD

**Discussion by Kcko** [tailwindlabs/tailwindcss#7896] [Help]
Hi,

Im using tailwind (last version) + scss together via post-css.
Everything works fine. 

Im using npm script for watch mode
``` "dev:tw:watch2": "set TAILWIND_MODE=watch&postcss --env dev ./www/sass/screen.scss -o ./www/assets/css/app.dev.css --watch --verbose",``` and its works fine.

But Tailwind doesn't track my files with a `.latte extension` (similar to a twig system). I must resave css / scss file and TW regenerate all ...  its confusing.

**tailwind.config.js**
===
```
module.exports = {
    //content: ['./src/**/*.{html,js}'],
    content: ['**/*.latte'],
    theme: {
        extend: {},
    },
    plugins: [],
};

```

And here is short video hot it works / dont works ;-)

https://bit.ly/3ua5n3F

Can you help me  how to set it right or what am I doing wrong?

**Comment by Kcko** [comment]
I tried with .html and same result, watch mode works me only with css file .. :(

**Comment by Kcko** [comment]
Hmm strange!

With concrete file it fully works!  
test.latte is on the same level like **tailwind.config.js**
` content: ['./app/**/*.{html,latte}', './test.latte'],`

But i really have correct structure ( `app/FrontModule/templates/RealEstate/rozcestnik.latte`) so `'./app/**/*.{html,latte}'` should it work right?

Video:  https://bit.ly/3IoI1fK


I don't understand and I'm starting to get unhappy 😣

Edit: with full path works again 
`content: ['./app/**/*.{html,latte}', './test.latte', './app/FrontModule/templates/RealEstate/rozcestnik.latte'],`

**Comment by Kcko** [comment]
its look like that failing with wildcars, but why

**Comment by adamwathan** [comment]
Hey! Can you please provide a proper reproduction project? Impossible to help without seeing all of the code.

I’m not sure if postcss-cli supports the newer PostCSS directory dependency API yet, it was behind for a while. I don’t think you should be using PostCSS to compile Sass though, which is why it would be really helpful to see a full reproduction, because it sounds like you have a lot of stuff going on here and the source of the problem is almost definitely not something you’ve provided here so far.

**Reply by Kcko** [reply]
Hi Adam, at first thanks for fast reply and your time generally.

Here is my cutted repo https://github.com/Kcko/tw-sass-watchmode (without php scripts & php vendors etc), just Tailwind + Sass + all config files (package.json, babel etc).

I need use tailwind together with sass due to advanced things like  import, nesting (both very essential) and sass extra (loops, conditionals, mixins ...)

I didn't come up with any other way how to connect TW + SASS than postCss,  if you recommend me a normal solution, I will be happy (not type Webpack etc) 

I hope that you find a problem with wildcards ;-), because  this solution (TW+SASS) works well ...

**Reply by adamwathan** [reply]
Taking a quick look now — what is the actual script you run to build your CSS? Do you run the gulp scripts, then the PostCSS scripts?

I would expect that the way this works is you first compile the Sass to CSS, then you feed that CSS into PostCSS, but I can't quite figure out what I'm looking at here.

From the `postcss-scss` docs:

<img width="738" alt="image" src="https://user-images.githubusercontent.com/4323180/159165949-6f7ce252-c2a3-4adb-a17c-16d06ada30b8.png">

That library _doesn't actually compile Sass_, it just understands the syntax and lets you process the AST.

You can get imports and nesting without using Sass:

https://tailwindcss.com/docs/using-with-preprocessors#using-post-css-as-your-preprocessor

The reason I say all this is because my gut is that the root problem here isn't in Tailwind itself but is coming from some the complexity of how everything is interacting with such a complicated setup.

I see you are using Parcel to compile your JS for example — why not also use Parcel to compile your Sass and PostCSS stuff? It has support for that built right in. You could also try Laravel Mix which has built-in support for compiling Sass and piping it through PostCSS 👍

I would start by just really trying to simplify things here and see if you still have the issue. The most glaring thing to me is the `postcss-scss` stuff which I don't think you actually need because I don't think it does what you need — you need something separate to compile .scss to .css, then feed that .css into a PostCSS pipeline.

Hope that helps in some way!

**Reply by Kcko** [reply]
Hi Adam,

im using `dev:tw:watch2`

**Reply by Kcko** [reply]
Im trying parcel for bundling & compiling JS files (from ES 6 => ES 5 and concatening => onle bundled file).  I haven't tried it with CSS yet, but i can later.

Try with `npm run dev:tw:watch2` and make some change in test.latte (same depth level like tailwind.config.js) => will work, but if you do it in app\front.... it wont, watch mode is sleeping (wildcards not works)

**Comment by Kcko** [comment]
Hi, any progress ...

Folder structure
- App
  - FrontModule
    - templates
      - Homepage
        - default.latte
        

if I try the whole way at the first time ` content: ['./app/FrontModule/templates/Homepage/default.latte']` it works!
        
**Other wilcards options already work, but only for that file, not for others**
```
 content: ['./app/**/**/*.latte'],
 content: ['./app/FrontModule/**/*.latte'],
 content: ['./app/FrontModule/templates/**/*.latte'],
 content: ['./app/FrontModule/**/*.latte'],
 content: ['./app/**/**/**/default.latte'],
 content: ['./app/**/**/**/*.latte']
 ```

I still think there's a problem with TW or Wildcards, not how it's compiled. Everything is compiling well, it's just the fuc**** 😢  watch mode and its recognition.

I'll jump off the rock soon

**Comment by Kcko** [comment]
I switched to parcel.

**.postcssrc**
```
{
  "plugins": {
    "postcss-import": {},
    "tailwindcss/nesting": {},
    "tailwindcss": {},
    "autoprefixer": {},
    "cssnano": {}
  }
}
```
**npm command**
```
"watch-css": "parcel watch ./www/assets/css/app.scss --dist-dir public/css --public-url /css"
`` 

tailwind.config.js
`content: ['./app/**/*.latte'],` 


**WORKS!**

Can you tell me it my solution is ok?
Something suitable to add, SCSS import and nested stuff work too

:-)

**Comment by jsmrtn** [comment]
Hey @adamwathan, I'm also seeing this issue in a different environment.

Here's the basic setup you can play with https://joshuamartin-builddemo-bf5h6t9oars.ws-eu54.gitpod.io/

If you run `npm run dev`, it does correctly pick up styles from all of the views, the JS files, and the safelist. The gulp task also correctly creates the example CSS file with the Tailwind classes rendered.

However, if you edit `system/views/frontend/guide/home.view` the JIT works fine, if you edit `system/views/frontend/guide/product.view` it doesnt run JIT. The former is hardcoded in `tailwind.config.js` and the latter is globbed. If I edit anything in the JS files in `src/**/*.js` it also doesn't run the JIT compiler to add any classes from the JS into `layout.css` (though it will also run the gulp task to run browserify on the files to put them into dist). Adding anything to `.safelist` works all the time.

**Reply by jsmrtn** [reply]
rtfm amirite. I've re-engaged with it this morning and updated my postCSS command from `TAILWIND_MODE=watch postcss ./src/css/style.css -o ./public_html/dist/layout.css -w --verbose` to `tailwindcss -i ./src/css/style.css -o ./public_html/dist/layout.css --watch --verbose` and it's now tracking globbed files correctly.
